### DISCUSSION TITLE
CORS error in S3 File Upload

### DISCUSSION THREAD

**Discussion by czarnndz** [Budibase/budibase#6324] [Help] [answered]
Budibase Cloud 1.0.199 

I'm trying to use a S3 File Upload component in Form, but I'm getting CORS error

![image](https://user-images.githubusercontent.com/7936586/173729642-3e208e1e-a268-4c3c-951a-1df7633854ff.png)

I set S3 Source without Endpoint, also I tried with and without Región but I saw that Request always use eu-west-1, maybe that could be the error idk.
![image](https://user-images.githubusercontent.com/7936586/173729751-3feb3cac-287c-43d8-bf80-9e7f25f3afaf.png)

This are the config I'm using
![image](https://user-images.githubusercontent.com/7936586/173730511-b15fbb20-f6e9-4f67-aa18-96a6028903d6.png)

And I'm using the "Upload File to S3" Action in "On change" event
![image](https://user-images.githubusercontent.com/7936586/173730642-96226f28-3e86-4b37-8ef2-238b84b7a24f.png)

The AWS User I'm using has administrator policy,
Below is the CORS bucket configuration in S3 I'm using

`[
    {
        "AllowedHeaders": [
            "*"
        ],
        "AllowedMethods": [
            "PUT",
            "POST",
            "DELETE",
            "GET"
        ],
        "AllowedOrigins": [
            "https://hic.budibase.app"
        ],
        "ExposeHeaders": []
    },
    {
        "AllowedHeaders": [
            "*"
        ],
        "AllowedMethods": [
            "PUT",
            "POST",
            "DELETE",
            "GET"
        ],
        "AllowedOrigins": [
            "*"
        ],
        "ExposeHeaders": []
    }
]`

Hope that someone can help me with this.

Thanks and Regards

**Comment by melohagan** [comment, marked as answer]
Hi @czarnndz 

I was able to replicate this bug for your selected region and CORS config supplied.

> I set S3 Source without Endpoint, also I tried with and without Región but I saw that Request always use eu-west-1, maybe that could be the error idk.

I think you are onto something. This does work for **eu-west-1** if you are able to use a bucket created for that region instead in the meantime?
I've raised a bug here: https://github.com/Budibase/budibase/issues/6332

**Reply by czarnndz** [reply]
Hi @melohagan, 

I create a new bucket in eu-west-1 and now it works perfectly
![image](https://user-images.githubusercontent.com/7936586/173843291-efa82caf-6551-4deb-b235-fb7744d78ac0.png)

Thank you so much for your help!!!

**Reply by melohagan** [reply]
You're very welcome. I have a PR out for a fix; might be a bit before it gets released.

**Comment by sergey-vin** [comment]
hi @czarnndz , which field type you used for `profile_img_s3`? I use postgresql as a DB and tried `text` or `text[]` field / column types and it always said "This Field setting is the wrong data type for this component"

**Reply by melohagan** [reply]
Hey @sergey-vin I imagine it's the *Attachment* field type of the internal Budibase DB. Attachments are currently not supported in Postgres, there's an issue open here: https://github.com/Budibase/budibase/issues/5486

**Reply by sergey-vin** [reply]
i tried Attachment for BBDB, didn't work as well. Says "This Field setting is the wrong data type for this component". I couldn't find any use to the s3 upload component unfortunately.

**Reply by melohagan** [reply]
You can always give it a unique name. It doesn't have to match a field on your table.

**Reply by melohagan** [reply]
I've been working on some improvements for the S3 connector so maybe that would make the upload component more useful in your apps when it's released: https://github.com/Budibase/budibase/pull/7521

**Reply by sergey-vin** [reply]
ok, so the expected use case is to give it a fake field name, then add an action which uploads to s3? If so, then i tried it and it didn't upload anything again, but then more questions have to be asked:
- if this "s3 upload component" is supposed to replace an "attachment" field, then how can I mimic a "gallery" behavior of an "attachment" field, when I can upload 2+ photos, and all of them will be saved in an array? S3 upload component has a "file name" field, which assumes that it's only 1 file, not a collection
- from one of the tickets here i found that s3 upload is a one-way street - you can only upload files, you can't display the content if the file was already uploaded. That makes s3 upload somewhat useless - i won't know if something is already uploaded, and can accidentally re-upload/rewrite old content.

I guess my overall impression is that this "s3 upload" is not a replacement to "attachments", and there's no way to upload any pictures or files to any non-internal DB. Am I right?

**Reply by melohagan** [reply]
> I guess my overall impression is that this "s3 upload" is not a replacement to "attachments", and there's no way to upload any pictures or files to any non-internal DB. Am I right?

I think this summary is spot on. S3 Upload allows you to upload a file to an S3 bucket. If the file was an image, then you could use the upload URL as the source for an image component. 
CSV files may be able to be read as well, but otherwise I think the main use case is simply if a user wants to upload a file that can then be read at a later point by some external application.

**Reply by sergey-vin** [reply]
I see. so a potential implementation of Attachments for external DBs would be:
* have a data provider -> repeater -> image/card linked to a DB table with multiple records, 1 record = 1 gallery image
* have a "delete button" near each card/image to drop them
* have an s3 upload component to upload new image to s3, and add a new row to the DB table with images, then refresh data provider

wow. that's quite some work. better to have support for Attachments in external DBs (potentially, using S3 as a storage for images and store URLs in DB only)
